### PR TITLE
[WIP] Use session cookie to identify current account instead of url param

### PR DESCRIPTION
**Pull request by danielfriis**

### CODE REVIEW

**Comment by danielfriis** [comment]
@kaspernj Curious to hear what you think of this :) Only implemented for templates#index

File: app/helpers/account_switches_helper.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+module AccountSwitchesHelper
+  def link_to_account_switch(account)
+    link_to account.name, account_switches_path(account_id: account.id), method: :post
+  end
+end
```
Comment (kaspernj) [inline app/helpers/account_switches_helper.rb]: Missing newline

File: app/helpers/account_switches_helper.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+module AccountSwitchesHelper
+  def link_to_account_switch(account)
+    link_to account.name, account_switches_path(account_id: account.id), method: :post
```
Comment (kaspernj) [inline app/helpers/account_switches_helper.rb]: You can simplify you path calls with arrays, if you like:
```ruby
link_to account.name, [account, :switches], method: :post
```

File: spec/rails_helper.rb [hunk 3]
```diff
@@ -20,7 +20,7 @@
 # directory. Alternatively, in the individual `*_spec.rb` files, manually
 # require only the support files necessary.
 #
-# Dir[Rails.root.join('spec/support/**/*.rb')].each { |f| require f }
+Dir[Rails.root.join('spec/support/**/*.rb')].each { |f| require f }
```
Comment (kaspernj) [inline spec/rails_helper.rb]: I believe Rubocop is set to double quotes

**Comment by kaspernj** [comment]
@danielfriis My opinion here is that doing it with sessions is a bad idea, because it wont be clear from the URL, which account a given resource belong to. If you change account, and then send me a link to a resource, then the account will be the default one for me, and so the resource shouldn't exist.

You can of course hack yourself out of that, by detecting which account a given resource belongs to and then redirect the user... But that is a hack. And you would probably have to implement that for all the models.

Account based on domain name, like on Advokurser, would be different, because that changes a part of the URL - the domain name.

**Comment by danielfriis** [comment]
@kaspernj It's a valid point. But what if I really don't want account to be part of the URL?

**Comment by kaspernj** [comment]
@danielfriis Maybe as a subdomain? That would also enable you to scale your application through more physical machines in the future (if it gets insanely popular)

**Comment by danielfriis** [comment]
But what if I really really dont't want it to be part of the URL? :D

**Comment by danielfriis** [comment]
I see it almost in the same way as you don't have `www.example.com/users/1/resource/1`, but instead have `www.example.com/resource/1`

**Comment by kaspernj** [comment]
@danielfriis Ok. But then why have `account` in the session at all?

What would be shown on your example, if I accessed `/resources`? All the resources or the resources for the users that is set in my session variable?

What if I chose another user that wasn't mine (user-2) for my session, and I then accessed `/resources/1`where resource-1 belonged to user-1, but the user in the session was user-2?

**Comment by danielfriis** [comment]
Then CanCan would raise a unauthorized exception. Just as if you were logged in as user-1, trying to access ´/account/2´ owned by user-2

**Comment by kaspernj** [comment]
@danielfriis Yes exactly. So any links to resources would assume a specific value in your session, which would cause a lot of those links to fail.

Links and URL's shouldn't be dependent on anything else than their own value. No session specific values in your session as an example.

We are also talking SEO here. If Google picked up a link somewhere, it would try and access it to index the content. It would then get a `NotFound` or `AccessDenied` error, because it didn't have the right session value present. I assume it would then give up. Even though this isn't a concern in the current project, I believe it can be regarded as bad practice.

If somebody wrote a manual and linked a specific budget for his organisation, that link would again assume a specific value in the session to be present. Anyone following that link without that value would again experience an error.

It would make it very problematic to be a member of more than one organisation. It would at least make it very problematic to use links from outside the app itself. Even bookmarks would fail when having the wrong organisation ID in the session variable.

**Comment by danielfriis** [comment]
@kaspernj Argh, I hate when I'm arguing against valid and great points! :D Maybe the solution for my OCD is shorting the URL? `/a/1/templates` eg.

**Comment by kaspernj** [comment]
@danielfriis I cannot think of a better solution either :-/ The only thing I would have against that, is that the URL's aren't that pretty. But I guess that is the lesser of two evils.

**Comment by danielfriis** [comment]
@kaspernj, if I wanted a public sharable pretty and short URL to a given resource. Could (would) I then just set a new route for that action and resource in the routes file?

**Pull request closed without merging**